File: lb1/java/WorkerThreadsController.java
import java.util.ArrayList;
import java.util.List;
import java.util.Random;

class WorkerThreadsController extends Thread {
    private List<ThreadConfiguration> _threads;
    private boolean _isFinished = false;

    public synchronized boolean isFinished() {
        return this._isFinished;
    }

    private synchronized void setIsFinished(boolean value) {
        this._isFinished = value;
    }

    public WorkerThreadsController(int threadsAmount) {
        this._threads = this.generateThreads(threadsAmount);
        this.run();
    }

    @Override
    public void run() {
        long now = System.currentTimeMillis();

        this._threads.forEach((entry) -> {
            entry.thread.start();
        });

        while (!this._isFinished) {
            this.setIsFinished(true);

            this._threads.forEach((entry) -> {
                if (entry.thread.isAlive()) {
                    this.setIsFinished(false);

                    if (!entry.thread.isInterrupted()) {
                        long elapsed = System.currentTimeMillis() - now;
                        if (elapsed >= entry.timeoutMs) {
                            entry.thread.interrupt();
                        }
                    }
                }
            });
        }
    }

    private List<ThreadConfiguration> generateThreads(int count) {
        Random random = new Random();
        List<ThreadConfiguration> result = new ArrayList<ThreadConfiguration>();
        int min = 20;
        int max = 50;

        for (int i = 0; i < count; i++) {
            int runDuration = random.nextInt(max - min + 1) + min;
            double step = (random.nextInt(max - min + 1) + min) * 10.0;
            result.add(new ThreadConfiguration(new WorkerThread(i, step), runDuration * 1000));
        }

        return result;
    }
}

File: lb1/c#/lb1c#/WorkerThreadsController.cs

namespace lb1 {
    class WorkerThreadsController {
        private List<ThreadConfiguration> _threads;
        private bool _isFinished = false;

        public bool IsFinished
        {
            get { lock(this) { return this._isFinished; }}
            private set { lock(this) { this._isFinished = value; }}
        }

        public WorkerThreadsController(int threadsAmount) {
            this._threads = this.GenerateThreads(threadsAmount);
        }

        public void Run() {
            new Thread(() => {
                long now = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();

                this._threads.ForEach((entry) => {
                    entry.thread.Start();
                });

                while (!this.IsFinished) {
                    this.IsFinished = true;

                    this._threads.ForEach((entry) => {
                        if (entry.thread.IsAlive) {
                            this.IsFinished = false;

                            long elapsed = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds() - now;
                            if (elapsed >= entry.timeoutMs && !entry.thread.IsInterrupted) {
                                // System.Console.WriteLine(entry.timeoutMs);
                                entry.thread.Interrupt();
                            }
                        }
                    });
                }
            }).Start();
        }

        private List<ThreadConfiguration> GenerateThreads(int count) {
            var random = new Random();
            var result = new List<ThreadConfiguration>();
            int min = 1;
            int max = 10;

            for (int i = 0; i < count; i++)
            {
                Random rnd = new Random();
                int runDuration = rnd.Next(max - min + 1) + min;
                double step = (rnd.Next(max - min + 1) + min) * 10;
                result.Add(new ThreadConfiguration(new WorkerThread(i, step), runDuration * 1000));
            }

            return result;
        }
    }
}